File: src/sources/php/parallel.php
<?php

function worker(int $id, array $jobs, array &$results): void
{
    foreach ($jobs as $job) {
        echo "Worker $id started job $job\n";
        Fiber::suspend();
        echo "Worker $id finished job $job\n";
        $results[] = $job * 2;
    }
}

$jobs = [1, 2, 3, 4, 5];
$results = [];

$fibers = [];
foreach (range(1, 3) as $id) {
    $fibers[] = new Fiber(fn () => worker($id, $jobs, $results));
}

foreach ($fibers as $fiber) {
    $fiber->start();
}

while (count($results) < count($jobs)) {
    foreach ($fibers as $fiber) {
        if ($fiber->isSuspended()) {
            $fiber->resume();
        }
    }
}

print_r($results);

File: src/sources/python/parallel.py
import asyncio

async def worker(id, jobs, results):
    while not jobs.empty():
        job = await jobs.get()
        print(f"Worker {id} started job {job}")
        await asyncio.sleep(1)
        print(f"Worker {id} finished job {job}")
        results.append(job * 2)

async def main():
    jobs = asyncio.Queue()
    results = []

    for job in range(1, 6):
        await jobs.put(job)

    workers = [asyncio.create_task(worker(i, jobs, results)) for i in range(1, 4)]

    await asyncio.gather(*workers)

    print(results)

asyncio.run(main())